### PR TITLE
feat: add recursive inserstion sort

### PR DESCRIPTION
**Pull request by Fandroid745**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: insertion_sort_recursive.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    tests();
+
+    /// For user interaction
+    size_t n;
+    std::cout << "Enter the length of your array (0 to exit): ";
+    std::cin >> n;
+    if (n == 0) {
+        return 0;
+    }
+
```
Comment (realstealthninja) [inline insertion_sort_recursive.cpp:133]: please remove user interaction from the main function! main function must be empty except for the call to tests and return 0

File: insertion_sort_recursive.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        arr[j + 1] = arr[j];
+        j--;
+    }
+    arr[j + 1] = last;
+}
+
+/** Recursive Insertion Sort Function for a vector
+ *
+ * @tparam T type of vector elements
+ * @param [in,out] arr pointer to vector to be sorted
```
Comment (realstealthninja) [inline insertion_sort_recursive.cpp:55]: n param should be documented

File: insertion_sort_recursive.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+/** \namespace sorting
+ * \brief Sorting algorithms
+ */
+namespace sorting {
+/** \brief
```
Comment (realstealthninja) [inline insertion_sort_recursive.cpp:22]: please use @ sign for your tags

File: insertion_sort_recursive.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+ * (Insertion Sort)](https://en.wikipedia.org/wiki/Insertion_sort)
+ *
+ * \details
+ * Insertion sort is a simple sorting algorithm that builds the final
+ * sorted array one at a time. The recursive version applies the same
+ * logic but sorts the sub-array recursively before inserting the current
+ * element in its correct position.
+ */
+
+#include <algorithm>
```
Comment (realstealthninja) [inline insertion_sort_recursive.cpp:13]: includes should be documented

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: sorting/insertion_sort.cpp [hunk 5]
```diff
@@ -1,101 +1,90 @@
 /**
```
Comment (realstealthninja) [inline sorting/insertion_sort.cpp:1]: you've added two files please remove one of them

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: sorting/insertion_sort.cpp [hunk 6]
```diff
@@ -1,101 +1,90 @@
...
- * the algorithms goes in the manner that we start iterating over the array
- * of elements as soon as we find a unsorted element that is a misplaced
- * element we place it at a sorted position.
- *
- * Example execution steps:
- * 1. Suppose initially we have
- * \f{bmatrix}{4 &3 &2 &5 &1\f}
- * 2. We start traversing from 4 till we reach 1
- * when we reach at 3 we find that it is misplaced so we take 3 and place
- * it at a correct position thus the array will become
```
Comment (realstealthninja) [inline sorting/insertion_sort.cpp:31]: this is a modification of a pre-existing algorithm. Please create a new file with the recursive suffix added to it

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: sorting/insertion_sort_recursive.cpp [hunk 7]
```diff
@@ -0,0 +1,150 @@
+/**
+ * \file
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:2]: please use \@ file & other \@ tags instead of \\tag

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: sorting/insertion_sort_recursive.cpp [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+ * 1. Start with the array [4, 3, 2, 5, 1].
+ * 2. Insert 3 in its correct position: [3, 4, 2, 5, 1].
+ * 3. Insert 2: [2, 3, 4, 5, 1].
+ * 4. Continue this until the array is sorted: [1, 2, 3, 4, 5].
+ */
+
+#include <algorithm>   ///< for std::is_sorted
+#include <cassert>     ///< for assert function in testing
+#include <iostream>    ///< for std::cout and std::endl
+#include <vector>      ///< for using std::vector
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:30]:
```suggestion
#include <algorithm>   /// for std::is_sorted
#include <cassert>     /// for assert function in testing
#include <iostream>    /// for std::cout and std::endl
#include <vector>      /// for using std::vector
```

File: sorting/insertion_sort_recursive.cpp [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+ * @namespace sorting
+ * @brief Contains sorting algorithms
+ */
+namespace sorting {
+
+/**
+ * @brief Insertion Sort Function
+ * 
+ * @tparam T Type of the array elements
+ * @param [in,out] arr Array to be sorted
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:42]:
```suggestion
 * @param[in,out] arr Array to be sorted
```

File: sorting/insertion_sort_recursive.cpp [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    assert(std::is_sorted(arr6, arr6 + 50));
+    std::cout << "passed" << std::endl;
+}
+
+/** 
+ * @brief Main function
+ * 
+ * Empty except for the call to `tests()`.
+ * 
+ * @return 0 Always returns 0.
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:144]:
```suggestion
 * @return 0 on sucessful exit
```

File: sorting/insertion_sort_recursive.cpp [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+ */
+int main() {
+    /// Running predefined tests to test the algorithm
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:147]:
```suggestion
```

File: sorting/insertion_sort_recursive.cpp [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    tests();
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:148]:
```suggestion
    tests(); // run self test implementations
```

File: sorting/insertion_sort_recursive.cpp [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    while (N--) {
+        double r = (std::rand() % 10000 - 5000) / 100.f;
+        arr[N] = static_cast<T>(r);
+    }
+}
+
+/** 
+ * @brief Test Cases for the sorting algorithm 
+ */
+void tests() {
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:99]:
```suggestion
static void tests() {
```

File: sorting/insertion_sort_recursive.cpp [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+ * - Stable: it does not change the relative order of elements with equal keys.
+ * 
+ * Insertion sort works similarly to how people sort playing cards in their hands.
+ * The algorithm iterates through the list and inserts each element into its correct 
+ * position in the sorted portion of the array.
+ *
+ * Example execution:
+ * 1. Start with the array [4, 3, 2, 5, 1].
+ * 2. Insert 3 in its correct position: [3, 4, 2, 5, 1].
+ * 3. Insert 2: [2, 3, 4, 5, 1].
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:23]: Please add references/sources for further reading

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: sorting/insertion_sort_recursive.cpp (hunk 14 above)
Comment (Fandroid745) [reply sorting/insertion_sort_recursive.cpp:27]: Added the requested changes you asked for.
My contribution is for Hacktober fest.

File: sorting/insertion_sort_recursive.cpp [hunk 15]
```diff
@@ -0,0 +1,156 @@
+/**
+ * @file
+ * @brief Insertion Sort Algorithm
+ * @author [Dhanush S] (https://github.com/Fandroid745)
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:4]:
```suggestion
 * @author [Dhanush S](https://github.com/Fandroid745)
```

File: sorting/insertion_sort_recursive.cpp [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+template <typename T>
+static void create_random_array(T *arr, int N) {
+    while (N--) {
+        double r = (std::rand() % 10000 - 5000) / 100.f;
+        arr[N] = static_cast<T>(r);
+    }
+}
+
+/** 
+ * @brief Test Cases for the sorting algorithm
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:102]:
```suggestion
 * @brief self test implementation
 * @return void
```

File: sorting/insertion_sort_recursive.cpp [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    sorting::insertionSort(arr6, 50);
+    assert(std::is_sorted(arr6, arr6 + 50));
+    std::cout << "passed" << std::endl;
+}
+
+/** 
+ * @brief Main function
+ * 
+ * Empty except for the call to `tests()`.
+ *
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:148]:
```suggestion
```

File: sorting/insertion_sort_recursive.cpp [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+/** 
+ * @brief Main function
+ * 
+ * Empty except for the call to `tests()`.
+ * 
+ * @return 0 on successful exit.
+ */
+int main() {
+    tests(); /// run self test implementations
+
```
Comment (realstealthninja) [inline sorting/insertion_sort_recursive.cpp:153]:
```suggestion
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by realstealthninja** [comment]
Hey we are close to approving this. Question is this for hacktoberfest?

**Comment by Fandroid745** [comment]
> Hey we are close to approving this. Question is this for hacktoberfest? 
Yes

**Review by realstealthninja — APPROVED** [review]
Looks good to me!

**Pull request merged**